File: typescript/decimal_to_binary.ts
// Decimal to binary number using recursion and iteration
import { tramp, Trampoline } from './global';

function decimalToBinary(num: number): string {
    let binary = '';
    while (num > 0) {
        const remainder = num % 2;
        num = Math.floor(num / 2);
        binary = remainder + binary;
    }
    return binary;
}

function decimalToBinaryRec(num: number): string {
    if (num == 0) {
        return ''
    }
    const remainder = num % 2;
    num = Math.floor(num / 2);
    return decimalToBinaryRec(num) + remainder;
}

function decimalToBinaryTailRec(num: number): string {
    function helper(num: number, binary = ''): string {
        if (num == 0) {
            return binary;
        }
        const remainder = num % 2;
        num = Math.floor(num / 2);
        return helper(num, remainder + binary);
    }
    return helper(num);
}

function decimalToBinaryThunk(num: bigint): Trampoline<string> {
    function helper(num: bigint, binary = ''): Trampoline<string> {
        if (num == 0n) {
            return binary;
        }
        const remainder = num % 2n;
        num = num / 2n;
        return () => helper(num, remainder + binary);
    }
    return helper(num);
}

const num = 101321009;

console.log(num.toString(2));
console.log(decimalToBinary(num));
console.log(decimalToBinaryRec(num));
console.log(decimalToBinaryTailRec(num));
//console.log(decimalToBinaryTailRec(2 ** 10_000));
console.log(tramp(decimalToBinaryThunk, 2n ** 10_000n));

File: python/src/decimal_to_binary.py
# Decimal to binary number using recursion and iteration
from common import time_execution


@time_execution()
def decimal_to_binary(num: int) -> str:
    res = ""
    while num > 0:
        remainder = num % 2
        num = num // 2
        res = str(remainder) + res
    return res


@time_execution()
def decimal_to_binary_rec(num: int) -> str:
    def helper(num: int) -> str:
        if num == 0:
            return ""
        return helper(num//2)+str(num % 2)
    return helper(num)


@time_execution()
def decimal_to_binary_tail_rec(num: int) -> str:
    def helper(num: int, binary_str="") -> str:
        if num == 0:
            return binary_str
        remainder = str(num % 2)
        return helper(num//2, remainder+binary_str)
    return helper(num)


@time_execution(isTrampoline=True)
def decimal_to_binary_gen(num: int) -> str:
    def helper(num: int, binary_str=""):
        if num == 0:
            yield binary_str
        remainder = str(num % 2)
        yield helper(num//2, remainder+binary_str)
    return helper(num)


print(decimal_to_binary(8))
print(decimal_to_binary_rec(8))
print(decimal_to_binary_tail_rec(8))
# print(decimal_to_binary_tail_rec(2**1000))
print(decimal_to_binary_gen(2**1000))
